### ISSUE TITLE
Broken character literal highlighting

### ISSUE THREAD

**Issue by nhatanh002** [neovimhaskell/haskell-vim#132]
When a character literal contains a "region" start pattern ('(','[','{','"'), everything after the start pattern is interpreted as a contained region, and consequently breaks syntax highlighting. After further investigation, the regex used to match character literal is actually incorrect according to the Haskell 2010 language specification (https://www.haskell.org/onlinereport/haskell2010/haskellch2.html#x7-200002.6), which means most valid character literals aren't highlighted at all, and the \u\<code\> pattern while not being a legal character in Haskell, is correctly highlighted.

For example, '[' here breaks everything after it: ) is not highlighted as a delimiter -  like its matching (, and showVol is not highlighted as it should've been, like getVol above:
![2021-11-20T01:31:47,889252436+07:00](https://user-images.githubusercontent.com/5835891/142677214-3dc726b4-aa4b-44e5-ba93-2ebb69cc8630.png)

With the addition of a ' and a ], they "close" the bogus "open" token [ and ', and everything is highlighted correctly again:
![2021-11-20T01:33:18,134189766+07:00](https://user-images.githubusercontent.com/5835891/142677212-59d44906-1c26-4146-a671-5c4deafac5ca.png)

The character literal highlighting is not even correct!
![2021-11-20T01:47:43,282991015+07:00](https://user-images.githubusercontent.com/5835891/142677207-d47429cc-d643-425b-b867-934833a558fe.png)

This is how the plugin defines the way haskellCharacter should've been highlighted, notice that the "characters" inside of the single quotes are highlighted as if they *weren't* character literals instead, e.g. A is highlighted as a type, 9 as a number, + as an operator, z as normal text:
![2021-11-20T01:49:22,307889855+07:00](https://user-images.githubusercontent.com/5835891/142677202-8bc44b29-a5f1-4510-9d79-ba9bced0960a.png)

**Comment by d86leader** [comment]
Oh god it's not just me. The character literals used to work fine, then I took some time off haskell, and now they're broken for both vim and neovim.

**Comment by d86leader** [comment]
Yep, just tested it and it's a regression somewhere between neovim 0.4.1 and neovim 0.6.0. It's got something to do with matching `\<'`, and it's the reason I couldn't fix your MR of this issue. Going to open an issue in neovim.
